tempesta_fw/tls.c

```diff
@@ -19,73 +19,583 @@
+		if (r == TFW_BLOCK)
+			return TFW_BLOCK;
+
+		return TFW_PASS;
+	}
+
+	tls_err(c, "-- handshake failed (%x)", -r);
+
+	return TFW_BLOCK;
+}
+
+
+/**
+ * Send @buf of length @len using TLS context @tls.
+ */
+static inline int
+tfw_tls_send_buf(TfwConnection *c, const unsigned char *buf, size_t len)
+{
+	int r;
+	TfwTlsContext *tls = c->tls;
+
+	tls_dbg(c, "=>");
+
+	while ((r = ttls_ssl_write(tls, buf, len)) > 0) {
+		if (r == len)
+			return 0;
+		buf += r;
+		len -= r;
+	}
+
+	tls_err(c, "-- write failed (%x)", -r);
+
+	return -EINVAL;
 }
 
-/* TODO alloc TLS- or HTTPS-specific message. */
-static TfwMsg *
-tfw_tls_conn_msg_alloc(TfwConnection *conn)
+/**
+ * Send @skb using TLS context @tls.
+ */
+static inline int
+tfw_tls_send_skb(TfwConnection *c, struct sk_buff *skb)
 {
-	return NULL;
+	int i;
+
+	tls_dbg(c, "=>");
+
+	if (skb_headlen(skb)) {
+		if (tfw_tls_send_buf(c, skb->data, skb_headlen(skb)))
+		    return -EINVAL;
+	}
+
+	for (i = 0; i < skb_shinfo(skb)->nr_frags; i++) {
+		const skb_frag_t *f = &skb_shinfo(skb)->frags[i];
+		if (tfw_tls_send_buf(c, skb_frag_address(f), f->size))
+		    return -EINVAL;
+	}
+
+	kfree_skb(skb);
+
+	return 0;
 }
 
-/* TODO Update @conn for newly established connection */
+/**
+ * Callback function which is called by TLS library.
+ */
 static int
-tfw_tls_conn_init(TfwConnection *conn)
+tfw_tls_send_cb(void *conn, const unsigned char *buf, size_t len)
 {
+	TfwConnection *c = conn;
+	TfwTlsContext *tls = c->tls;
+	struct sk_buff *skb;
+
+	tls_dbg(c, "=>");
+
+	skb = alloc_skb(MAX_TCP_HEADER + len, GFP_ATOMIC);
+	if (unlikely(!skb))
+		return MBEDTLS_ERR_NET_SEND_FAILED;
+
+	skb_reserve(skb, MAX_TCP_HEADER);
+	skb_put(skb, len);
+
+	if (unlikely(skb_store_bits(skb, 0, buf, len)))
+		BUG();
+
+	ss_skb_queue_tail(&tls->tx_queue, skb);
+	if (ss_send(c->sk, &tls->tx_queue, 0))
+		return MBEDTLS_ERR_NET_SEND_FAILED;
+
+	tls_dbg(c, "-- %lu bytes sent", len);
+
+	return len;
+}
+
+/**
+ * Callback function which is called by TLS library.
+ */
+static int
+tfw_tls_recv_cb(void *conn, unsigned char *buf, size_t len)
+{
+	TfwConnection *c = conn;
+	TfwTlsContext *tls = c->tls;
+	struct sk_buff *skb = ss_skb_peek_tail(&tls->rx_queue);
+
+	tls_dbg(c, "=>");
+
+	if (unlikely(!skb))
+		return 0;
+
+	len = min_t(size_t, skb->len, len);
+	if (unlikely(skb_copy_bits(skb, 0, buf, len)))
+		BUG();
+
+	pskb_pull(skb, len);
+
+	if (unlikely(!skb->len)) {
+		ss_skb_unlink(&tls->rx_queue, skb);
+		kfree_skb(skb);
+	}
+
+	tls_dbg(c, "-- %lu bytes rcvd", len);
+
+	return len;
+}
+
+static void
+tfw_tls_conn_dtor(TfwConnection *c)
+{
+	TfwTlsContext *tls = c->tls;
+
+	mbedtls_ssl_free(&tls->ssl);
+
+	kmem_cache_free(tfw_tls_cache, tls);
+	c->tls = NULL;
+
+	tfw_cli_conn_release(c);
+}
+
+static int
+tfw_tls_conn_init(TfwConnection *c)
+{
+	int r;
+	TfwTlsContext *tls;
+
+	tls_dbg(c, "=>");
+
+	BUG_ON(c->tls);
+	tls = c->tls = kmem_cache_alloc(tfw_tls_cache, GFP_ATOMIC);
```

Comment:
Each TLS connection has `TfwTlsContext` as well as `TfwConnection`. So why to do tow SLAB calls? There is no reason to do so. Please make TLS connection inherited from TfwConnection or make more elegant (`tls` pointing to the end of function isn't such) way if you can invent such.